### PR TITLE
Refactor core interfaces and directory structure

### PR DESCRIPTION
**Pull request by skipor**
Ideas:
* Put each component in it's directory in `components`.
* Put core packages in `core` directory.
* Put utils packages in `lib` directory
* More type safety in components
* Make `Aggregator` (previous `ResultListener`) abstract
* Move component imports to its `import` directories

### CODE REVIEW

File: components/phttp/connect.go [hunk 1]
```diff
@@ -123,7 +124,7 @@ func newConnectDialer(target string, connectSSL bool, dialer Dialer) Dialer {
 		}
 		// No need to close body.
 		if r.Buffered() != 0 {
-			// Already receive something non HTTP from proxy or dialed server.
+			// Already receive something non Ammo from proxy or dialed server.
```
Comment (direvius) [inline components/phttp/connect.go]: Auto-replacement bug

File: components/phttp/connect.go [hunk 2]
```diff
@@ -89,7 +90,7 @@ func newConnectDialer(target string, connectSSL bool, dialer Dialer) Dialer {
 		req := &http.Request{
 			Method:     "CONNECT",
 			URL:        &url.URL{},
-			Proto:      "HTTP/1.1",
+			Proto:      "Ammo/1.1",
```
Comment (direvius) [inline components/phttp/connect.go]: Auto-replacement bug

File: components/phttp/connect.go [hunk 3]
```diff
@@ -10,13 +10,14 @@ import (
...
-	"github.com/yandex/pandora/gun"
+
+	"github.com/yandex/pandora/core"
 )
 
 type ConnectGunConfig struct {
 	Target     string       `validate:"endpoint,required"`
 	ConnectSSL bool         `config:"connect-ssl"` // Defines if tunnel encrypted.
-	SSL        bool         // As in HTTP gun, defines scheme for http requests.
+	SSL        bool         // As in Ammo gun, defines scheme for http requests.
```
Comment (direvius) [inline components/phttp/connect.go]: Auto-replacement bug

File: components/phttp/ammo/simple/ammo.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+	req *http.Request
+	tag string
+}
+
+func NewAmmo(req *http.Request, tag string) *Ammo {
+	return &Ammo{req, tag}
+}
+
+func (a *Ammo) Request() (*http.Request, *netsample.Sample) {
+	sample := netsample.Acquire(a.tag)
```
Comment (direvius) [inline components/phttp/ammo/simple/ammo.go]: Acquiring ammo and acquiring measurement instrument should not be logically coupled.

Sample should be acquired in a shoot method, not here. Sometimes we need more then one sample per one call, sometimes we don't want any samples to be sent. I want this code to be ready for writing custom http guns with complex scenarios over it.

File: core/engine/engine.go [hunk 5]
```diff
@@ -36,107 +34,99 @@ func (e *Engine) Serve(ctx context.Context) error {
...
+	for ctx.Err() == nil {
+		// Acquire should unblock in case of context cancel.
+		ammo, more := u.Provider.Acquire()
+		if !more {
+			log.Println("Ammo ended")
+			break
+		}
+		_, more = <-control
+		if !more {
+			log.Println("Schedule ended.")
```
Comment (skipor) [inline core/engine/engine.go]: fixed

File: components/phttp/connect.go (hunk 3 above)
Comment (skipor) [inline components/phttp/connect.go]: fixed

File: components/phttp/ammo/simple/ammo.go (hunk 4 above)
Comment (skipor) [inline components/phttp/ammo/simple/ammo.go]: Let's change that logic in separate PR.

**Pull request merged**